File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-5.cpp
class Solution {
public:
    int characterReplacement(string s, int k) {
        unordered_map<char, int> charCounts;
        
        int maxCount = 0;

        int left = 0, maxFrequency = 0;

        for (int right = 0; right < s.size(); ++right) {
            ++charCounts[s[right]];
            maxFrequency = max(maxFrequency, charCounts[s[right]]);

            while ((right - left + 1 ) - maxFrequency > k) {
                --charCounts[s[left]];
                ++left;
            }

            maxCount = max(maxCount, right - left + 1);
        }

        return maxCount;
    }
};

File: Data Structures & Algorithms/burst-balloons/submission-5.py
class Solution:
    def maxCoins(self, nums: List[int]) -> int:
        numsPad = [1]
        numsPad.extend(nums)
        numsPad.append(1)

        n = len(numsPad)
        dp = [[0 for _ in range(n)] for _ in range(n)]

        def dfs(left, right):
            if left > right:
               return 0 
            elif left == right:
                return numsPad[left - 1] * numsPad[left] * numsPad[left + 1]
            elif dp[left][right] > 0:
                return dp[left][right]

            maxValue = 0
            for value in range(left, right + 1):
                leftSide = dfs(left, value - 1)
                rightSide = dfs(value + 1, right)

                popValue = numsPad[left - 1] * numsPad[value] * numsPad[right + 1]
                maxValue = max(popValue + leftSide + rightSide, maxValue)

            dp[left][right] = maxValue
            return dp[left][right]

        return dfs(1, len(numsPad) - 2)
                    
